src/execution/riscv64/simulator-riscv64.cc

```diff
@@ -3035,7 +3038,9 @@ void Simulator::DecodeRVSType() {
 #endif /*V8_TARGET_ARCH_64_BIT*/
     // TODO: use F Extension macro block
     case RO_FSW: {
-      WriteMem<float>(rs1() + s_imm12(), frs2(), instr_.instr());
+      WriteMem<uint32_t>(rs1() + s_imm12(), 
+                         (uint32_t)get_fpu_register_word(rs2_reg()), 
+                         instr_.instr());
```

Comment:
This is confusing. Why not implement FSW as a floating-point write? I think it can be done as
```
WriteMem<float>(sr1() + s_imm12(),  get_fpu_register_float(rs2_reg()), instr_.instr());
```

Reply:
The handling of `FLW` needs to nan-box the loaded value.

> Floating-point n-bit transfer operations move external values held in IEEE standard formats into
and out of the f registers, and **comprise floating-point loads and stores (FLn/FSn)** and floatingpoint
move instructions (FMV.n.X/FMV.X.n). **A narrower n-bit transfer, n < FLEN, into the f
registers will create a valid NaN-boxed value**. A narrower n-bit transfer out of the floating-point
registers will transfer the lower n bits of the register ignoring the upper FLEN−n bits.

```
case RO_FLW: {
      int64_t addr = rs1() + imm12();
      float val = ReadMem<float>(addr, instr_.instr());
      set_frd(val, false); <== need to be nan-boxed
      TraceMemRd(addr, val, get_fpu_register(frd_reg()));
      break;
    }
```

Reply:
Before I explain the questions about FLW and FSW, I'd like to explain some things related.

`box float` and `check boxed float` will be done in `set_fpu_register_float` and `get_fpu_register_float`.

```
void Simulator::set_fpu_register_float(int fpureg, float value) {
  DCHECK((fpureg >= 0) && (fpureg < kNumFPURegisters));
  FPUregisters_[fpureg] = box_float(value);
}

float Simulator::get_fpu_register_float(int fpureg) const {
  DCHECK((fpureg >= 0) && (fpureg < kNumFPURegisters));
  if (!is_boxed_float(FPUregisters_[fpureg])) {
    return std::numeric_limits<float>::quiet_NaN();
  }
  return *bit_cast<float*>(const_cast<int64_t*>(&FPUregisters_[fpureg]));
}
```

Back to the questions.

The original implementation of FSW with `frs2` which invokes `get_fpu_register_float`. But the handling of `FSW` should not `check boxed float`.

> A narrower n-bit transfer out of the floating-point
registers will transfer the lower n bits of the register ignoring the upper FLEN−n bits.

As for `FLW`, it's no problem. Because `set_frd` would invoke `set_fpu_register_float` which boxes float number.

```
inline void set_frd(float value, bool trace = true) {
    set_fpu_register_float(rd_reg(), value);
    if (trace) TraceRegWr(get_fpu_register_word(rd_reg()), FLOAT);
  }
```